lib/plaid/http_client.ex

```diff
@@ -0,0 +1,84 @@
+defmodule Plaid.HTTPClient do
+  @moduledoc """
+  HTTP Client behaviour default implementation.
+  """
+
+  defmodule Response do
+    @derive Jason.Encoder
+    defstruct status_code: nil, body: nil
+    @type t :: %__MODULE__{status_code: integer, body: term}
+  end
+
+  defmodule Error do
+    @derive Jason.Encoder
+    defexception reason: nil, message: nil
+    @type t :: %__MODULE__{reason: term}
+    @impl true
+    def message(%Error{reason: reason, message: nil}),
+      do: "http request failed with reason: #{inspect(reason)}"
+
+    def message(%Error{reason: nil, message: message}), do: message
+  end
+
+  @type method :: atom
+  @type url :: String.t()
+  @type body :: map
+  @type headers :: keyword
+  @type http_options :: keyword
+  @type metadata :: map
+  @type response :: {:ok, Plaid.HTTPClient.Response.t()} | {:error, Plaid.HTTPClient.Error.t()}
+
+  @doc """
+  Sends HTTP request using default configuration.
+
+  Default HTTP Client implementation using Tesla. Function accepts request
+  arguments: `method`, `url`, `body`, `headers`, HTTP client configuration options,
+  and metadata for telemetry and then sends the request.
+
+  The function must return either `{:ok, %Plaid.HTTPClient.Response{}}`
+  or `{:error, %Plaid.HTTPClient.Error{}}`.
+
+  Default implementation uses a Tesla client with the following options:
+  ```
+  adapter = {Tesla.Adapter.Hackney, []}
+  middleware = [Tesla.Middleware.JSON, Plaid.Telemetry]
+  ```
+
+  Users can modify this implementation by creating their own module which
+  implements the `Plaid.HTTPClient` behaviour and setting it in their library
+  configuration under `:http_client` or passing it in the configuration
+  argument at runtime.
+  """
+  @callback call(method, url, body, headers, http_options, metadata) :: response
+
+  @doc false
+  def call(method, url, body, headers, http_options \\ [], metadata \\ %{}) do
+    client = build_client(http_options)
+
+    options = [
+      method: method,
+      url: url,
+      headers: headers,
+      body: body,
+      opts: [metadata: metadata]
+    ]
+
+    case Tesla.request(client, options) do
+      {:ok, %Tesla.Env{status: status, body: body}} ->
+        {:ok, %Plaid.HTTPClient.Response{status_code: status, body: Jason.decode!(body)}}
+
+      {:error, reason} ->
+        {:error, %Plaid.HTTPClient.Error{reason: reason}}
```

Comment:
What would be the status code in the error branching?

I am using it today to return with such status code back to the client that called my backend.

Reply:
The HTTP request fails in the error branch so no status code is generated. You can see this use case in the following test.
```
$mix test test/lib/plaid_test.exs:323
```
Status codes `>299` will be returned by the `Plaid.HTTPClient` in the `{:ok, Plaid.HTTPClient.Response.t}` branch.

However, the `Plaid` module isn't designed to be called directly. If you were fetching accounts, for example, you would call `Plaid.Accounts.get/2` and it would return either `{:ok, %Plaid.Accounts{}}` or `{:error, %Plaid.Error{}}` both of which are successful HTTP requests, but the former has a status code `200..299` and the latter `>299`.

If for some reason the HTTP response failed, `Plaid.Accounts.get/2` would return `{:error, %Plaid.HTTPClient.Error{}}` with the reason. So you can distinguish between an invalid call to Plaid and a failed HTTP request.

Reply:
This particular return value would be the equivalent of [`result()`](https://hexdocs.pm/tesla/Tesla.Env.html#t:result/0) type in Tesla. If the HTTP request succeeds, it will return a response in the format of `{:ok, Tesla.Env.t}` with some status code, 200-500. If the HTTP request fails completely it will return `{:error, any()}`. I don't believe any of the Tesla adapters return a status code in that case. Hackney returns an atom, such as `:econnrefused`.

However, your question is making me more certain that I need to drop this `Plaid.HTTPClient` abstraction because it is just obfuscating the return values you would expect by mapping `{:ok, Tesla.Env.t} -> {:ok, Plaid.HTTPClient.Response.t}` and `{:error, any()} -> {:error, Plaid.HTTPClient.Error}`. There should only be four cases a user of this library should need to deal with:

1. Raise - User fails to configure or provide credentials needed for an API call to Plaid
2. Success - HTTP request succeeds and returns the desired structured data. Can match on `{:ok, %Plaid.Accounts{} | %Plaid.Transactions{} | ... }`
3. Error - HTTP request succeeds and returns an Plaid error (https://plaid.com/docs/errors/#errors-overview), should always match on `{:error, %Plaid.Error{}}`
4. Failure - HTTP request fails due to closed connection, timeout, etc. This is the raw response by the configured adapter, which in most cases is passed through by Tesla in the format `{:error, any()}` In this branch, I was mapping this case to `{:error, %Plaid.HTTPClient.Error{}}`, which is confusing.

Reply:
I love your finding, really clear details in terms of what we should do, also I learned today a good abstraction, in general, to write these clients and deal with the Tesla error vs client error.

> Raise - User fails to configure or provide credentials needed for an API call to Plaid

We could do that `only` when creating the client instead of per-requests if we drop the usage of client id and client secret from the body and move that configuration to be part of the header only.

I am not sure what would be the downside of such a decision though. I still don't know why Plaid allows both places to add such a thing or why we would use the body one. Worst case, it could add it once somebody asks for it.

Here is my implementation of it, I don't show fancy messages but can be added

```elixir
defmodule Plaid.Client do
  alias Plaid.Client.Request

  def new do
    Tesla.client([
      {Tesla.Middleware.BaseUrl, base_url()},
      {Tesla.Middleware.Headers,
       [
         {"user-agent", "Elixir"},
         {"Plaid-Version", "2020-09-14"},
         {"PLAID-CLIENT-ID", client_id()},
         {"PLAID-SECRET", client_secret()}
       ]},
      Tesla.Middleware.JSON
    ])
  end

  def send_request(request, client) do
    request
    |> Request.to_keyword()
    |> then(&Tesla.request!(client, &1))
  end

  defp client_id do
    Keyword.fetch!(app_env(), :client_id)
  end

  defp client_secret do
    Keyword.fetch!(app_env(), :client_secret)
  end

  defp base_url do
    case Keyword.fetch!(app_env(), :base_url) do
      :sandbox -> "https://sandbox.plaid.com"
      :production -> "https://production.plaid.com"
      :development -> "https://development.plaid.com"
    end
  end

  defp app_env do
    Application.get_env(:plaid_client, __MODULE__, [])
  end
end
```

Reply:
> We could do that only when creating the client instead of per-requests if we drop the usage of client id and client secret from the body and move that configuration to be part of the header only.

I really like this design approach and I think it achieves the same result. The only potential downside I can see is that it's a little more awkward to _not_ submit credentials for endpoints that don't need them, but that only applies to the `/categories` endpoint in Plaid (and I don't know if Plaid ignores them or errors if they're provided).

This is how I was thinking about implementing.
```elixir
# lib/client/plaid.ex
defmodule Plaid.Client do

  defmodule Request do
    defstruct [body: nil, endpoint: nil, method: nil, opts: %{}]

    @spec to_options(Request.t) :: keyword
    def to_options(%Request{body: b, endpoint: e, method: m, opts: o}) do
      [method: m, url: e, body: b, opts: Map.to_list(o)]
    end

    @spec put_metadata(Request.t, map) :: Request.t
    def put_metdata(%Request{body: b, endpoint: e, method: m, opts: o} = request, config) do
      metadata = 
        Map.new()
        |> Map.put(:method, m)
        |> Map.put(:path, e)
        |> Map.put(:u, :native)
        |> Map.merge(config[:telemetry_metadata] || %{})

      %{request | opts: Map.merge(o, metadata}
    end
  end

  @spec new(map) :: Tesla.Client.t
  def new(config \\ %{}) do
    middleware = [
      {Tesla.Middleware.BaseUrl, get_base_url(config)},
      {Tesla.Middleware.Headers,
       [
        {"Content-Type", "application/json"},
        {"user-agent", "Elixir-SDK"},
        {"Plaid-Version", "2020-09-14"},
        {"PLAID-CLIENT-ID", get_client_id(config)},
        {"PLAID-SECRET", get_secret(config)}
       ]},
      Tesla.Middleware.JSON,
      Plaid.Telemetry
    ]

    adapter = {get_adapter(config), get_http_options(config)}

    Tesla.client(middleware, adapter)
  end

  defp get_base_url(config) do
    # reads url from the config -> app env -> raises if not found
  end

  defp get_client_id(config), do: # same thing
  
  defp get_secret(config), do: # same thing
  
  defp get_http_options(config) do
    # merges runtime http options provided, such as [recv_timeout: 30_000], into app env options
  end

  defp get_adapter(config) do
    config[:adapter] || Application.get_env(:plaid, :adapter, Tesla.Adapter.Hackney)
  end
end
```
**Rationale:** Creating a Tesla client from the config preserves the runtime configuration in the current version of the library. This also centralizes the common business logic for all HTTP requests and takes advantage of Tesla's capabilities.

Responsibilities for testing:

- Handling validation of credentials either passed in `config` argument or set in app env
- Handling runtime HTTP options provided to adapter
- Handling correct base url environment
- Handling correct middleware
- Translating request to the correct Tesla options for HTTP request
- Handling generation of default telemetry metadata and custom metadata provided

`Plaid.Client` and `Plaid.Client.Request` could be used in the following way:
```elixir
# lib/plaid.ex
defmodule Plaid do
  alias Plaid.Client.Request

  @callback send_request(Request.t, Tesla.Client.t) :: {:ok, Tesla.Env.t} | {:error, any()}

  @callback handle_response({:ok, Tesla.Env.t} | {:error, any()}) :: {:ok, term} | {:error, Plaid.Error.t} | {:error, any()}

  @doc false
  def send_request(request, client) do
    request
    |> Request.to_options()
    |> then(&Tesla.request(client, &1))
  end
 
  @doc false
  def handle_response({:ok, %Tesla.Env{status: status} = env}) when status in 200..299 do
    # pass through for decoding by calling function
    {:ok, env.body}
  end

  def handle_response({:ok, %Tesla.Env{} = env}) do
   # unmarshal response body into structured data
   {:error, %Plaid.Error{}}
  end

  def handle_response({:error, _reason} = error) do
    error
  end

end
```
**Rationale:** Functionality for submitting a request and handling the response. I implemented this as a behaviour so it can make testing easier to setup through mocking. For example, while Tesla has built-in mocking functions, I'd rather my unit test for `Plaid.Accounts.get/2` assert that the `%Request{}` value submitted to `Plaid.send_request/2` is correct than validate it downstream by mocking `Tesla.Client`.

Responsibilities for testing:

- Making the HTTP request with the correct options set
- Handling the main branches of the HTTP response, 1) success, 2) HTTP success but Plaid error, 3) HTTP failure

This would be called in the following way:
```elixir
# lib/plaid/accounts.ex
defmodule Plaid.Accounts do
  alias Plaid.Client
  alias Plaid.Client.Request

  defstruct accounts: [], item: nil, request_id: nil

  @endpoint "accounts"
  
  def get(params, config \\ %{}) do
    Request
    |> struct([body: params, method: :post, endpoint: "#{@endpoint}/get"])
    |> Request.put_metadata(config)
    |> Plaid.send_request(Client.new(config))
    |> Plaid.handle_response()
    |> case do
      {:ok, body} ->
        # unmarshal body to structured data (need to figure out how to do this using Poison or home-rolled)
        {:ok, %Plaid.Accounts{}}
      {:error, _} = error ->
        error
    end
  end
end
```
**Rationale:** This is a more declarative structure as you suggested: build the request, submit the request to the client, handle the response. This structure is runtime configurable, backwards compatible, and relocates the business logic of unmarshalling the response body to the module that owns the structured data.

Responsibilities for testing:

- Submitting the correct request parameters
- Handling custom metadata
- Mapping the success response from Plaid to the module's structured data

Finally, an application might use the library in the following way:
```elixir
defmodule Web.Api.Controller do
   use Web, :controller

   def list_accounts(conn, _params) do
       token = ...

        case Plaid.Accounts.get(%{access_token: token}, %{http_options: [recv_timeout: 10_000]}) do
            {:ok, %Plaid.Accounts{} = resp_body} ->
                conn
                |> put_status(200)
                |> json(resp_body)
            {:error, %Plaid.Error{} = error} ->
                conn
                |> put_status(400)
                |> json(%{message: error.display_message})
            {:error, reason} ->
                # special handling of HTTP failure if desired
        end
   end
end
```

Reply:
This might work for the decoding to structured data piece https://github.com/IvanRublev/Nestru

Reply:
> Tesla.Middleware.PathParams

It seems to be a good middleware as well, even closer to codegen stuff.